go-controller/pkg/ovn/gateway/gateway.go

```diff
@@ -87,3 +87,21 @@ func CreateDummyGWMacBindings(nbClient libovsdbclient.Client, nodeName string) e
 
 	return nil
 }
+
+// DeleteDummyGWMacBindings removes mac bindings (ipv4 and ipv6) for a fake next hops
+// used by host->service traffic
+func DeleteDummyGWMacBindings(nbClient libovsdbclient.Client, nodeName string) error {
```

Comment:
northd is not cleaning up the mac binding for us when the logical port is gone, this might be a northd bug though

Reply:
@dceara @numansiddique is this a bug?

Reply:
The schema for NB. (static) mac bindings is very relaxed:
```
        "Static_MAC_Binding": {
            "columns": {
                "logical_port": {"type": "string"},
                "ip": {"type": "string"},
                "mac": {"type": "string"},
                "override_dynamic_mac": {"type": "boolean"}},
            "indexes": [["logical_port", "ip"]],
             "isRoot": true},
```

The schema for the SB mac binding table is very similar.  The only difference is that ovn-northd doesn't expect anyone else to write to the SB directly (ovn-kube did but that's not really the right way to interact with OVN) so ovn-northd is in charge of creating/deleting SB.MAC_Bindings.

For the NB, it's a bit different, ovn-northd didn't create the NB.Static_MAC_Binding record, it's not clear whether it should clean it or not.

There's always the possibility that the CMS/user/admin/etc creates the Static_MAC_Binding before creating the LRP.  The schema allows that.  If northd were to automatically delete all "stale" Static_MAC_Bindings, this use case would break.

TL/DR: I don't really think this is a bug.  NB records are usually not owned by ovn-northd.

Reply:
@dceara 

While the NB records are usually not owned by ovn-northd, we would have a different behavior with the referential integrity that could have happened if `logical_port` was a strong reference and `Static_MAC_Binding` was a non root table.

I am not sure if there is a real use case behind the current schema design or if this has been perhaps an oversight when copying the SBDB table definition to NBDB which might have more implications.

One of those implications, and probably the real bug, is that northd fails to reconcile if `logical_port` does not exist which probably should be handled more gracefully with the current schema design.

Reply:
It's not OK to add the strong reference now.  We'd potentially break users.  I suspect there are users.  This feature was developed upstream and on a community member request: https://github.com/ovn-org/ovn/commit/b22684d4e37c3dd205a959925b6e3f1827bbe045


> One of those implications, and probably the real bug, is that northd fails to reconcile if `logical_port` does not exist which probably should be handled more gracefully with the current schema design.

We cannot assume that the NB entry should be removed.  With the current schema the user is allowed to insert NB.Static_MAC_Binding and NB.Logical_Router_Port in any order they wish.